src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -807,253 +844,383 @@ RouterSelectTask(Query *query)
+	ShardInterval *shardInterval = NULL;
+
+	if (node == NULL)
 	{
 		return false;
 	}
-#endif
 
-	if (joinTree == NULL)
+	/* want to look at all RTEs, even in subqueries, CTEs and such */
+	if (IsA(node, Query))
 	{
-		return false;
+		return query_tree_walker((Query *) node, UpdateRelationNames, restrictionContext,
+								 QTW_EXAMINE_RTES);
 	}
 
-	quals = joinTree->quals;
-	if (quals == NULL)
+	if (!IsA(node, RangeTblEntry))
 	{
-		return false;
+		return expression_tree_walker(node, UpdateRelationNames, restrictionContext);
 	}
 
-	/* convert list of expressions into expression tree */
-	if (quals != NULL && IsA(quals, List))
+
+	newRte = (RangeTblEntry *) node;
+
+	if (newRte->rtekind != RTE_RELATION)
 	{
-		quals = (Node *) make_ands_explicit((List *) quals);
+		return false;
 	}
 
 	/*
-	 * Partition column must be used in a simple equality match check and it must be
-	 * place at top level conjustion operator.
+	 * Search for the restrictions associated with the RTE. There better be
+	 * some, otherwise this query wouldn't be elegible as a router query.
+	 *
+	 * FIXME: We should probably use a hashtable here, to do efficient
+	 * lookup.
 	 */
-	partitionColumnMatchExpression =
-		ColumnMatchExpressionAtTopLevelConjunction(quals, partitionColumn);
-
-	if (!partitionColumnMatchExpression)
+	foreach(relationRestrictionCell, restrictionContext->relationRestrictionList)
 	{
-		return false;
+		relationRestriction =
+			(RelationRestriction *) lfirst(relationRestrictionCell);
+
+		if (GetRTEIdentity(relationRestriction->rte) == GetRTEIdentity(newRte))
+		{
+			break;
+		}
+
+		relationRestriction = NULL;
 	}
 
-	/* make sure partition column is used only once in the query */
-	varClauseList = pull_var_clause_default(quals);
-	foreach(varClauseCell, varClauseList)
+	if (relationRestriction == NULL)
 	{
-		Var *column = (Var *) lfirst(varClauseCell);
-		if (equal(column, partitionColumn))
+		Relation relation = heap_open(newRte->relid, AccessShareLock);
+		TupleDesc tupleDescriptor = RelationGetDescr(relation);
+		int columnCount = tupleDescriptor->natts;
+		int columnIndex = 0;
+		List *valuesList = NIL;
+		List *collationList = NIL;
+
+		newRte->rtekind = RTE_VALUES;
+
+		for (columnIndex = 0; columnIndex < columnCount; columnIndex++)
 		{
-			partitionColumnReferenceCount++;
+			FormData_pg_attribute *attributeForm = tupleDescriptor->attrs[columnIndex];
+			if (!attributeForm->attisdropped)
+			{
+				Const *constValue = makeNullConst(attributeForm->atttypid,
+												  attributeForm->atttypmod,
+												  attributeForm->attcollation);
+				valuesList = lappend(valuesList, constValue);
+				collationList = lappend_oid(collationList, attributeForm->attcollation);
+			}
 		}
+
+
+		newRte->values_lists = NIL;
+		newRte->values_lists = lappend(newRte->values_lists, valuesList);
+		newRte->values_collations = collationList;
+		newRte->alias = copyObject(newRte->eref);
+		heap_close(relation, AccessShareLock);
```

Comment:
> clarification requested, we still call heap_close() right ? Because we are done with relation reference.

Correct. Just without releasing the lock (i.e. specify NoLock instead of
AccessShareLock). Otherwise a concurrent session could alter the table
during execution, making the previous checks potentially obsolete.